applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_structural_response_function.cpp

```diff
@@ -0,0 +1,942 @@
+    {
+        KRATOS_TRY;
+
+        this->Clear();
+        this->Check();
+
+        ModelPart& r_model_part = this->GetModelPart();
+        ModelPart& r_sensitivity_model_part = r_model_part.GetSubModelPart(mSensitivityModelPartName);
+
+        // Initialize flags.
+        #pragma omp parallel for
+        for (int i = 0; i< static_cast<int> (r_sensitivity_model_part.Nodes().size()); ++i) 
+        {
+            NodesContainerType::iterator it = r_sensitivity_model_part.NodesBegin() + i;
+            it->SetValue(UPDATE_SENSITIVITIES, true);
+        }
+        #pragma omp parallel for
+        for (int i = 0; i< static_cast<int> (r_sensitivity_model_part.Elements().size()); ++i) 
+        {
+            ElementsContainerType::iterator it = r_sensitivity_model_part.ElementsBegin() + i;
+            it->SetValue(UPDATE_SENSITIVITIES, true);
+        }
+        #pragma omp parallel for
+        for (int i = 0; i< static_cast<int> (r_sensitivity_model_part.Conditions().size()); ++i) 
+        {
+            ConditionsContainerType::iterator it = r_sensitivity_model_part.ConditionsBegin() + i;
+            it->SetValue(UPDATE_SENSITIVITIES, true);
+        }
+
+    if(mGradientMode == 1)
+    {
+        #pragma omp parallel for
+        for (int i = 0; i< static_cast<int> (r_model_part.Elements().size()); ++i) 
+        {
+            ElementsContainerType::iterator it = r_model_part.ElementsBegin() + i;
+            it->SetValue(DISTURBANCE_MEASURE, mDelta);
+        }
+        #pragma omp parallel for
+        for (int i = 0; i< static_cast<int> (r_model_part.Conditions().size()); ++i) 
+        {
+            ConditionsContainerType::iterator it = r_model_part.ConditionsBegin() + i;
+            it->SetValue(DISTURBANCE_MEASURE, mDelta);
+        }
+    }
+
+        KRATOS_CATCH("");
+    }
+
+    void AdjointStructuralResponseFunction::InitializeSolutionStep(){}
+
+    void AdjointStructuralResponseFunction::FinalizeSolutionStep()
+    {
+        KRATOS_TRY;
+
+        this->UpdateSensitivities();
+
+        KRATOS_CATCH("");
+    }
+
+    void AdjointStructuralResponseFunction::Check()
+    {
+        KRATOS_TRY;
+
+        ModelPart& r_model_part = this->GetModelPart();
+
+        KRATOS_ERROR_IF_NOT(r_model_part.HasSubModelPart(mSensitivityModelPartName))
+            << "No sub model part \"" << mSensitivityModelPartName << "\"" << std::endl;
+
+        KRATOS_CATCH("");
+    }
+
+    void AdjointStructuralResponseFunction::Clear()
+    {
+        KRATOS_TRY;
+
+        ModelPart& r_model_part = this->GetModelPart();
+
+        // Reset flags.
+        #pragma omp parallel for
+        for (int i = 0; i< static_cast<int> (r_model_part.Nodes().size()); ++i) 
+        {
+            NodesContainerType::iterator it = r_model_part.NodesBegin() + i;
+            it->SetValue(UPDATE_SENSITIVITIES, false);
+        }
+        #pragma omp parallel for
+        for (int i = 0; i< static_cast<int> (r_model_part.Elements().size()); ++i) 
```

Comment:
Use core VariableUtils for these loops. You can add the bool version if it's missing.